# Introduction to Pandas

## Introduction 

Pandas is a Python library used for working with data sets. It has functions for analysing, cleaning, exploring, and manipulating data. The name "Pandas" has a reference to both "Panel Data", and "Python Data Analysis" and was created by Wes McKinney in 2008.

Some advantages of Pandas are:
- Fast and efficient for manipulating and analyzing data
- Data from different file objects can be loaded
- Easy handling of missing data (represented as NaN)
- Size mutability: columns can be inserted and deleted from DataFrame and higher dimensional objects
- Data set merging and joining
- Flexible reshaping and pivoting of data sets
- Provides time-series functionality




## Getting started

If you have Python and Anaconda already installed on a system, then installation of Pandas is very easy.

Install it using this command:

To load the pandas package and start working with it, import the package as:

In [36]:
import pandas as pd


IPython.utils.text.SList

Pandas provide two data structures for manipulating data: Series and Data frames. We will focus on the second one in this introduction.

Pandas DataFrame is a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns). It consists of three principal components, the data, the rows, and the columns.

Let's load a data frame and inspect it:

In [38]:
import os
import json

#!pip install kaggle

!kaggle datasets list -s avocado --sort-by votes 

 

ref                                                    title                                     size  lastUpdated          downloadCount  voteCount  usabilityRating  
-----------------------------------------------------  ---------------------------------------  -----  -------------------  -------------  ---------  ---------------  
neuromusic/avocado-prices                              Avocado Prices                           629KB  2018-06-06 05:28:35         259368       3537  0.9705882        
moltean/fruits                                         Fruits 360                                 1GB  2021-09-12 19:19:48         126635       2774  0.875            
aelchimminut/fruits262                                 Fruits-262                                 6GB  2021-12-15 19:16:59           4055        118  0.875            
timmate/avocado-prices-2020                            Avocado Prices (2020)                      1MB  2021-02-26 18:18:17           8680         90  0.7058824 

In [42]:

f = open('/Users/aromanelli/Downloads/kaggle.json')
 
# returns JSON object as 
# a dictionary
data = json.load(f)
 
# Iterate through the json

for i in data:
    print([i])
 
f.close()

username = data['username']
key = data['key']

#os.environ["KAGGLE_USERNAME"] = username
#os.environ["KAGGLE_KEY"] = key

#!kaggle datasets download -q neuromusic/avocado-prices
#!unzip avocado-prices.zip  



['username']
['key']
Archive:  avocado-prices.zip
  inflating: avocado.csv             


In [44]:
avocado = pd.read_csv('/Users/aromanelli/Downloads/avocado.csv')
avocado.head() # print the first few rows

Unnamed: 0        Date  AveragePrice  Total Volume     4046       4225  \
0           0  2015-12-27          1.33      64236.62  1036.74   54454.85   
1           1  2015-12-20          1.35      54876.98   674.28   44638.81   
2           2  2015-12-13          0.93     118220.22   794.70  109149.67   
3           3  2015-12-06          1.08      78992.15  1132.00   71976.41   
4           4  2015-11-29          1.28      51039.60   941.48   43838.39   

     4770  Total Bags  Small Bags  Large Bags  XLarge Bags          type  \
0   48.16     8696.87     8603.62       93.25          0.0  conventional   
1   58.33     9505.56     9408.07       97.49          0.0  conventional   
2  130.50     8145.35     8042.21      103.14          0.0  conventional   
3   72.58     5811.16     5677.40      133.76          0.0  conventional   
4   75.78     6183.95     5986.26      197.69          0.0  conventional   

   year  region  
0  2015  Albany  
1  2015  Albany  
2  2015  Albany  
3  2015  Albany  
4  2015  Albany

In brief, the columns are:
- Date — The date of the sale
- AveragePrice — the average price of a single avocado
- Total Volume — Total number of avocados sold
- 4046 — Total number of avocados with PLU 4046 sold
- 4225 — Total number of avocados with PLU 4225 sold
- 4770 — Total number of avocados with PLU 4770 sold
- Total Bags — Total number of bags sold
- Small Bags — Number of small bags sold
- Large Bags — Number of large bags sold
- XLarge Bags — Number of extra-large bags sold
- type — conventional or organic
- year — the year of the sale
- region — the city or region of the observation

You can also create a data frame 'from scratch':

In [45]:
df_passengers = pd.DataFrame(
    {
        "Name": ["Braund, Mr. Owen Harris",
                 "Allen, Mr. William Henry",
                 "Bonnell, Miss. Elizabeth"],
        "Age": [22, 35, 58],
        "Sex": ["male", "male", "female"],
    }
)

df_passengers

Name  Age     Sex
0   Braund, Mr. Owen Harris   22    male
1  Allen, Mr. William Henry   35    male
2  Bonnell, Miss. Elizabeth   58  female

## Exercise
Different methods exist to inspect the dataframe and return statistical information. Try `.columns`, `index` (and not row!), ``.info()`` and ``.describe()`` on the `df` dataframe to answer the following questions:
- How many rows and columns are in this data frame?
- What type of data does the `year` column contains?
- What period of time does that table cover?
- What are the maximum and minimum average price?
- What is the mean of the number of bags sold per sale?

In [46]:
# Your code here

## Indexing, selecting

You can subset (a group of) individual columns:

In [48]:
# One individual column
df_passengers.Name
# or using the indexing ([]) operator:
df_passengers["Name"]

# Groups of column can be called as a list:
#df_passengers[["Name", "Age"]]


0     Braund, Mr. Owen Harris
1    Allen, Mr. William Henry
2    Bonnell, Miss. Elizabeth
Name: Name, dtype: object

Note that only the subset of the dataframe appears on your screen (the top 5 and bottom five values) for practical reasons.

A specific value can be called as:

In [49]:
df_passengers.Name[1]
# or
df_passengers["Name"][1]

'Allen, Mr. William Henry'

It is also possible to modify specific columns of create new columns:

In [50]:
df_passengers['Year of birth'] = 2023 - df_passengers['Age']
df_passengers

Name  Age     Sex  Year of birth
0   Braund, Mr. Owen Harris   22    male           2001
1  Allen, Mr. William Henry   35    male           1988
2  Bonnell, Miss. Elizabeth   58  female           1965

Individual rows can be selected based on the index, using `iloc`

In [51]:
df_passengers.iloc[0]
df_passengers.iloc[[0, 2]]

Name  Age     Sex  Year of birth
0   Braund, Mr. Owen Harris   22    male           2001
2  Bonnell, Miss. Elizabeth   58  female           1965

## Statistical information

The summary functions mentionned above can also be applied to individual columns. Try the folowing commands: 

In [52]:
print(df.year.describe())

print(df.AveragePrice.mean())

print(df.region.unique())

count    18249.000000
mean      2016.147899
std          0.939938
min       2015.000000
25%       2015.000000
50%       2016.000000
75%       2017.000000
max       2018.000000
Name: year, dtype: float64
1.405978409775878
['Albany' 'Atlanta' 'BaltimoreWashington' 'Boise' 'Boston'
 'BuffaloRochester' 'California' 'Charlotte' 'Chicago' 'CincinnatiDayton'
 'Columbus' 'DallasFtWorth' 'Denver' 'Detroit' 'GrandRapids' 'GreatLakes'
 'HarrisburgScranton' 'HartfordSpringfield' 'Houston' 'Indianapolis'
 'Jacksonville' 'LasVegas' 'LosAngeles' 'Louisville' 'MiamiFtLauderdale'
 'Midsouth' 'Nashville' 'NewOrleansMobile' 'NewYork' 'Northeast'
 'NorthernNewEngland' 'Orlando' 'Philadelphia' 'PhoenixTucson'
 'Pittsburgh' 'Plains' 'Portland' 'RaleighGreensboro' 'RichmondNorfolk'
 'Roanoke' 'Sacramento' 'SanDiego' 'SanFrancisco' 'Seattle'
 'SouthCarolina' 'SouthCentral' 'Southeast' 'Spokane' 'StLouis' 'Syracuse'
 'Tampa' 'TotalUS' 'West' 'WestTexNewMexico']


To compute the correlation coefficients between pairs of columns:

In [53]:
subdf=df[['AveragePrice', 'Total Volume', 'Total Bags']]  # selecting a subset of the inital df
subdf.corr()

AveragePrice  Total Volume  Total Bags
AveragePrice      1.000000     -0.192752   -0.177088
Total Volume     -0.192752      1.000000    0.963047
Total Bags       -0.177088      0.963047    1.000000

As expected, there is a tight correlation between the total number of avocados sold (`Total Volume`) and the number of bags sold (`Total Bags`). The average price decreases slightly with an increasing volume.

## Conditional selection

So far we've been indexing various strides of data, using structural properties of the DataFrame itself. To do interesting things with the data, however, we often need to ask questions based on conditions.

To select all the men in the above list:

In [54]:
df_passengers.Sex == 'male'

0     True
1     True
2    False
Name: Sex, dtype: bool

This operation produced a Series of True/False booleans based on the country of each record. This result can then be used inside of loc to select the relevant data:

In [55]:
df_passengers[df_passengers.Sex == 'male']

Name  Age   Sex  Year of birth
0   Braund, Mr. Owen Harris   22  male           2001
1  Allen, Mr. William Henry   35  male           1988

## Exercise

Add a column to the avocado dataframe indicating the total price for the sale.

Determine how many avocado sales made in the region of Atlanta have reached a price above $700000.

In [56]:
# Your code here

## Grouping and sorting

The `groupby` operation allows to group the data by the values in a specific column. 

We can then find for example the minimum average price per region, or per region and per year:

In [57]:
avocado.groupby('region').AveragePrice.min()

region
Albany                 0.85
Atlanta                0.62
BaltimoreWashington    0.95
Boise                  0.58
Boston                 0.85
BuffaloRochester       1.03
California             0.67
Charlotte              0.80
Chicago                0.70
CincinnatiDayton       0.44
Columbus               0.52
DallasFtWorth          0.65
Denver                 0.60
Detroit                0.48
GrandRapids            0.77
GreatLakes             0.73
HarrisburgScranton     0.91
HartfordSpringfield    0.86
Houston                0.51
Indianapolis           0.77
Jacksonville           0.54
LasVegas               0.54
LosAngeles             0.53
Louisville             0.56
MiamiFtLauderdale      0.59
Midsouth               0.91
Nashville              0.51
NewOrleansMobile       0.58
NewYork                0.77
Northeast              0.87
NorthernNewEngland     0.95
Orlando                0.58
Philadelphia           0.91
PhoenixTucson          0.46
Pittsburgh             0.87
Plains       

In [58]:
avocado.groupby(['region', 'year']).AveragePrice.min()

region            year
Albany            2015    0.93
                  2016    0.85
                  2017    1.13
                  2018    1.03
Atlanta           2015    0.93
                          ... 
West              2018    0.83
WestTexNewMexico  2015    0.68
                  2016    0.68
                  2017    0.52
                  2018    0.75
Name: AveragePrice, Length: 216, dtype: float64

The `sort_values()` method can be used to sort the data in many ways:

In [59]:
avocado.sort_values(by='AveragePrice')

Unnamed: 0        Date  AveragePrice  Total Volume        4046  \
15261          43  2017-03-05          0.44      64057.04      223.84   
7412           47  2017-02-05          0.46    2200550.27  1200632.86   
15473          43  2017-03-05          0.48      50890.73      717.57   
15262          44  2017-02-26          0.49      44024.03      252.79   
1716            0  2015-12-27          0.49    1137707.43   738314.80   
...           ...         ...           ...           ...         ...   
16720          18  2017-08-27          3.04      12656.32      419.06   
16055          42  2017-03-12          3.05       2068.26     1043.83   
14124           7  2016-11-06          3.12      19043.80     5898.49   
17428          37  2017-04-16          3.17       3018.56     1255.55   
14125           8  2016-10-30          3.25      16700.94     2325.93   

            4225      4770  Total Bags  Small Bags  Large Bags  XLarge Bags  \
15261    4748.88      0.00    59084.32      638.68    58445.64         0.00   
7412   531226.65  18324.93   450365.83   113752.17   330583.10      6030.56   
15473    4138.84      0.00    46034.32     1385.06    44649.26         0.00   
15262    4472.68      0.00    39298.56      600.00    38698.56         0.00   
1716   286858.37  11642.46   100891.80    70749.02    30142.78         0.00   
...          ...       ...         ...         ...         ...          ...   
16720    4851.90    145.09     7240.27     6960.97      279.30         0.00   
16055      77.36      0.00      947.07      926.67       20.40         0.00   
14124   10039.34      0.00     3105.97     3079.30       26.67         0.00   
17428      82.31      0.00     1680.70     1542.22      138.48         0.00   
14125   11142.85      0.00     3232.16     3232.16        0.00         0.00   

               type  year             region  
15261       organic  2017   CincinnatiDayton  
7412   conventional  2017      PhoenixTucson  
15473       organic  2017            Detroit  
15262       organic  2017   CincinnatiDayton  
1716   conventional  2015      PhoenixTucson  
...             ...   ...                ...  
16720       organic  2017  RaleighGreensboro  
16055       organic  2017  MiamiFtLauderdale  
14124       organic  2016       SanFrancisco  
17428       organic  2017              Tampa  
14125       organic  2016       SanFrancisco  

[18249 rows x 14 columns]

`sort_values()` defaults to an ascending sort, where the lowest values go first. However, sometimes we would prefer a descending order:

In [60]:
avocado.sort_values(by='AveragePrice', ascending=False)

Unnamed: 0        Date  AveragePrice  Total Volume        4046  \
14125           8  2016-10-30          3.25      16700.94     2325.93   
17428          37  2017-04-16          3.17       3018.56     1255.55   
14124           7  2016-11-06          3.12      19043.80     5898.49   
16055          42  2017-03-12          3.05       2068.26     1043.83   
16720          18  2017-08-27          3.04      12656.32      419.06   
...           ...         ...           ...           ...         ...   
1716            0  2015-12-27          0.49    1137707.43   738314.80   
15262          44  2017-02-26          0.49      44024.03      252.79   
15473          43  2017-03-05          0.48      50890.73      717.57   
7412           47  2017-02-05          0.46    2200550.27  1200632.86   
15261          43  2017-03-05          0.44      64057.04      223.84   

            4225      4770  Total Bags  Small Bags  Large Bags  XLarge Bags  \
14125   11142.85      0.00     3232.16     3232.16        0.00         0.00   
17428      82.31      0.00     1680.70     1542.22      138.48         0.00   
14124   10039.34      0.00     3105.97     3079.30       26.67         0.00   
16055      77.36      0.00      947.07      926.67       20.40         0.00   
16720    4851.90    145.09     7240.27     6960.97      279.30         0.00   
...          ...       ...         ...         ...         ...          ...   
1716   286858.37  11642.46   100891.80    70749.02    30142.78         0.00   
15262    4472.68      0.00    39298.56      600.00    38698.56         0.00   
15473    4138.84      0.00    46034.32     1385.06    44649.26         0.00   
7412   531226.65  18324.93   450365.83   113752.17   330583.10      6030.56   
15261    4748.88      0.00    59084.32      638.68    58445.64         0.00   

               type  year             region  
14125       organic  2016       SanFrancisco  
17428       organic  2017              Tampa  
14124       organic  2016       SanFrancisco  
16055       organic  2017  MiamiFtLauderdale  
16720       organic  2017  RaleighGreensboro  
...             ...   ...                ...  
1716   conventional  2015      PhoenixTucson  
15262       organic  2017   CincinnatiDayton  
15473       organic  2017            Detroit  
7412   conventional  2017      PhoenixTucson  
15261       organic  2017   CincinnatiDayton  

[18249 rows x 14 columns]

Finally, a dataframe can be sorted by more than one column at a time:

In [61]:
avocado.sort_values(by=['type','AveragePrice'])

Unnamed: 0        Date  AveragePrice  Total Volume        4046  \
7412           47  2017-02-05          0.46    2200550.27  1200632.86   
1716            0  2015-12-27          0.49    1137707.43   738314.80   
1752           36  2015-04-19          0.51    1366844.88  1097285.22   
4566           42  2016-03-06          0.51    1442973.47  1037699.01   
6622           52  2017-01-01          0.51    1475741.07   985040.19   
...           ...         ...           ...           ...         ...   
16720          18  2017-08-27          3.04      12656.32      419.06   
16055          42  2017-03-12          3.05       2068.26     1043.83   
14124           7  2016-11-06          3.12      19043.80     5898.49   
17428          37  2017-04-16          3.17       3018.56     1255.55   
14125           8  2016-10-30          3.25      16700.94     2325.93   

            4225      4770  Total Bags  Small Bags  Large Bags  XLarge Bags  \
7412   531226.65  18324.93   450365.83   113752.17   330583.10      6030.56   
1716   286858.37  11642.46   100891.80    70749.02    30142.78         0.00   
1752   164460.99   7534.30    97564.37    44646.67    52917.70         0.00   
4566   259846.68  14567.40   130860.38    76814.40    54045.98         0.00   
6622   290583.75   6313.88   193803.25    62497.42   131305.83         0.00   
...          ...       ...         ...         ...         ...          ...   
16720    4851.90    145.09     7240.27     6960.97      279.30         0.00   
16055      77.36      0.00      947.07      926.67       20.40         0.00   
14124   10039.34      0.00     3105.97     3079.30       26.67         0.00   
17428      82.31      0.00     1680.70     1542.22      138.48         0.00   
14125   11142.85      0.00     3232.16     3232.16        0.00         0.00   

               type  year             region  
7412   conventional  2017      PhoenixTucson  
1716   conventional  2015      PhoenixTucson  
1752   conventional  2015      PhoenixTucson  
4566   conventional  2016      PhoenixTucson  
6622   conventional  2017            Houston  
...             ...   ...                ...  
16720       organic  2017  RaleighGreensboro  
16055       organic  2017  MiamiFtLauderdale  
14124       organic  2016       SanFrancisco  
17428       organic  2017              Tampa  
14125       organic  2016       SanFrancisco  

[18249 rows x 14 columns]

## Exercise

Compare the mean average price for organic and conventional avocados.

In [ ]:
# Your code here

Create a new dataframe containing only the date of the sale, the average price and the region, sorted by date (from the most recent to the oldest one) and by total price (from the cheapeast to the most expensive).

In [ ]:
# Your code here

## Concatenation and Merging

**pandas** provides various methods for combining together Series or DataFrame with various kinds of logic.

The ``concat()`` function performs concatenation operations along an axis while performing optional set logic (union or intersection) of the indexes on the other axes.


In [67]:
df1 = pd.DataFrame(
    {
        "A": ["A0", "A1", "A2", "A3"],
        "B": ["B0", "B1", "B2", "B3"],
        "C": ["C0", "C1", "C2", "C3"],
        "D": ["D0", "D1", "D2", "D3"],
    },
    index=[0, 1, 2, 3],
)


df2 = pd.DataFrame(
    {
        "A": ["A4", "A5", "A6", "A7"],
        "B": ["B4", "B5", "B6", "B7"],
        "C": ["C4", "C5", "C6", "C7"],
        "D": ["D4", "D5", "D6", "D7"],
    },
    index=[4, 5, 6, 7],
)


df3 = pd.DataFrame(
    {
        "A": ["A8", "A9", "A10", "A11"],
        "B": ["B8", "B9", "B10", "B11"],
        "C": ["C8", "C9", "C10", "C11"],
        "D": ["D8", "D9", "D10", "D11"],
    },
    index=[8, 9, 10, 11],
)


frames = [df1, df2, df3]

result = pd.concat(frames, keys=["x", "y", "z"])


A    B    C    D
x 0    A0   B0   C0   D0
  1    A1   B1   C1   D1
  2    A2   B2   C2   D2
  3    A3   B3   C3   D3
y 4    A4   B4   C4   D4
  5    A5   B5   C5   D5
  6    A6   B6   C6   D6
  7    A7   B7   C7   D7
z 8    A8   B8   C8   D8
  9    A9   B9   C9   D9
  10  A10  B10  C10  D10
  11  A11  B11  C11  D11

How to handle the other axes ? 

This can be done in the following two ways:

- Take the union of them all, join='outer'. This is the default option as it results in zero information loss

- Take the intersection, join='inner'

In [84]:
df4 = pd.DataFrame(
    {
        "B": ["B2", "B3", "B6", "B7"],
        "D": ["D2", "D3", "D6", "D7"],
        "F": ["F2", "F3", "F6", "F7"],
    },
    index=[2, 3, 6, 7],
)
result = pd.concat([df1, df4], axis=1, join="inner")


A   B   C   D   B   D   F
2  A2  B2  C2  D2  B2  D2  F2
3  A3  B3  C3  D3  B3  D3  F3

**pandas** provides a single function, `merge()`, as the entry point for all standard database join operations between DataFrame objects:

![Alt text](image-2.png)

In [93]:
left = pd.DataFrame(
    {
        "key1": ["K0", "K0", "K1", "K2"],
        "key2": ["K0", "K1", "K0", "K1"],
        "A": ["A0", "A1", "A2", "A3"],
        "B": ["B0", "B1", "B2", "B3"],
    }
)


right = pd.DataFrame(
    {
        "key1": ["K0", "K1", "K1", "K2"],
        "key2": ["K0", "K0", "K0", "K0"],
        "C": ["C0", "C1", "C2", "C3"],
        "D": ["D0", "D1", "D2", "D3"],
    }
)

result = pd.merge(left, right, how="outer", on=["key1", "key2"])
result

key1 key2    A    B    C    D
0   K0   K0   A0   B0   C0   D0
1   K0   K1   A1   B1  NaN  NaN
2   K1   K0   A2   B2   C1   D1
3   K1   K0   A2   B2   C2   D2
4   K2   K1   A3   B3  NaN  NaN
5   K2   K0  NaN  NaN   C3   D3

## Exercise

Dowload the following dataset, using Kaggle API:

Name: data science for good kiva crowdfunding
Owner : kiva

Do some exploratory analysis on the loans.csv dataset:
- print the name of the columns and data types using built-in pandas functions
- extract some stats on the loans (counts, average loan, total amount funded), grouped by sector, country and both. 
- Add a column for the relative amount of loan funded wrt the total sum 
- Sort the values based on the total amount funded.

Use the merging method to glue different .csv files:

- do a join with the theme_ids.csv file, on the 'id' key and save the result of the join as a new dataframe
- join the result fo the previous point with the theme_by_region.csv file, this time using multiple keys: region, country, Loan Theme ID, Partner ID



## Windowing

**pandas** contains a set of methods for performing windowing operations - an operation that performs an aggregation over a sliding partition of values. The method functions similarly to the **groupby**  in that Series and DataFrame call the windowing method with necessary parameters and then subsequently call the aggregation function.

**pandas** supports 4 types of windowing operations:

- Rolling window: Generic fixed or variable sliding window over the values.
- Weighted window: Weighted, non-rectangular window 
- Expanding window: Accumulating window over the values.
- Exponentially Weighted window: Accumulating and exponentially weighted window over the values.

In [110]:
s = pd.Series(range(5), index=pd.date_range('2023-01-01', periods=5, freq='1D'))
s
s.rolling(window='2D').sum()


2023-01-01    0.0
2023-01-02    1.0
2023-01-03    3.0
2023-01-04    5.0
2023-01-05    7.0
Freq: D, dtype: float64


## Exercise 
Download this dataset:

`akashram/indian-summer-over-the-years`

After an Exploratory Data Analysis (EDA), use the window APIs to calculate and plot a median trend of the temperature measurements in India (see 'Median Filter' on Wikipedia).

The window method can be chained after a groupby. Make use of the chaining, to plot a median trend of the temperature per city. 

For the EDA: Pandas supports some APIs for plotting using Matplotlib. Try using `df.hist()` or `df.plot()`, along with additional parameters, to see how to properly and efficiently plot the columns of a Dataframe.